S7k/non tenant side (#169)
tenancy/docs: docs/tenancy/1.x/tutorial-non-tenant.md

---
title: Non-Tenant Site Setup
icon: fal fa-gears
excerpt: >
    How to setup goabal and admin sites in the same application
tags:
    - tenancy
    - tutorial
    - non-tenant
---

## Introduction
When building a SaaS there tends to be a need for a "sub-site" such as an admin panel, or a public facing website describing your SaaS.

This tutorial will go though the basics of setting these "sub-sites" within the same codebase without the need for creating a tenant do display these sites.

## Requirements
This tutorial will assume that you have followed the [basic setup tutorial](tutorial-basic-setup).

## Routes

> The first thing we want to do is configure the [Routes Affect](affacts-routes).

Reviewer: ```suggestion
The first thing we want to do is configure the [Routes Affect](affects-routes).
```

> First start by creating a new file for all tenant specific routes.

Reviewer: This is repeated from above. First we do, first we do that.

A more fluid way of saying this is something like

```suggestion
In order to set up the tenant specific routes, we will need to create a routes file for it.
```

### Tenant Route File
For this example it will be stored at `routes\tenant.php`
```php
use Illuminate\Support\Facades\Route;

/*
|--------------------------------------------------------------------------
| Tenant Web Routes
|--------------------------------------------------------------------------
|
| Here is where you can register web routes for your tenant application.
| They are not loaded by default, and will be loaded thought the Routes Affect.
|
*/

Route::get('/', function () {

>     //This is the "root" route for all tenants

Reviewer: I'd never use a callable in an example for routes. Callables cause routes files to be no longer cacheable. So.. either just remove the non-working example or provide a controlller from the get-go.

});

```

### Routes Affect Listener

> Next we need to create the listener for the [Routes Affect]() that will get rid of the system (non-tenant) routes and load our tenant route file instead.

Reviewer 2: Link to the right page please

`app\Listeners\ConfigureTenantRoutes.php`
```php
namespace App\Listeners;

use Tenancy\Affects\Routes\Events\ConfigureRoutes;

class TenantRoutes 
{
    public function handle(ConfigureRoutes $event) 
    {
        if($event->event->tenant)
        {
            $event
                ->flush()
                ->fromFile(
                    ['middleware' => ['web']],
                    base_path('/routes/tenant.php')
                );
        }
    }
}
```

#### Registering the listener
Now update your `app\Providers\EventServiceProvider.php` to include your new listener.

```php
  protected $listen = [

>     ...

Reviewer: ```suggestion
    // ...
```

    Tenancy\Affects\Routes\Events\ConfigureRoutes::class => [
      App\Listeners\ConfigureTenantRoutes::class,
    ]
    

>     ...

Reviewer: ```suggestion
    // ...
```

  ];
```

## Status

> At this point you application is configured to load any tenant routes when a tenant is identified. When a tenant is not identified, the routes located in `routes/web.php` will be used instead.

Reviewer: ```suggestion
At this point your application is configured to load any tenant routes when a tenant is identified. When a tenant is not identified, the routes located in `routes/web.php` will be used instead.
```

## Tips and Reminders
### Namespaceing
As with any Laravel application; if you are storing controllers in a different location than in `app\Http\Controllers` you will need to use the namespace option within your routes.

`routes\tenant.php`
```php
Route::namespace('Tenant')->group(function () {

>     // Controllers Within The "App\Http\Controllers\Tenant" Namespace

Reviewer: It's better to provide commented examples than commented explanations..

});
```

### Sub-site by domains
When using subdomains to identify your non-tenant sub-sites you will need to use the route option within your routes file.
`routes\web.php`
```php
Route::domain('admin.myapp.com')->group(function () {
    Route::get('/', function () {

>         // This is the root route for the admin site

Reviewer: No callables please.

    });
});

Route::domain('www.myapp.com')->group(function () {
    Route::get('/', function () {

>         // This is the root route for the public site

Reviewer: No callables.

    });

> });

Reviewer 2: Nope, no need :) You can do this with `identification-driver-http` rather than this way, it's kinda messed up imo.

Author: That implies that an admin panel should be created as a tenant.

"www.domain.com" and "admin.domain.com" are *not* tenants in this example.

I.E. if my admin panel is a tenant, how can I manage other tenants from within a tenant?

Route::get('/shared', function() {

>   //This is a route that is shared with both the admin site and the public site

Reviewer: No callables.

})
```

> ### Accessing `OnTenant` models from a "sub-site"

Reviewer: Can we rename "sub-site" to tenant site or something else? And make it consistent in the whole tutorial.

Before a model that uses the `OnTenant` trait can be used a tenant needs to be identified. Because of this it is best practice to avoid using tenant models on a non-tenant site.

> In events where this access is still needed, a tenant will first need to be identified within your code.

Reviewer: I don't even think this warning is necessary. Especially on the tenant website you will need to know who the tenant is.. In fact I would even enable eager tenant identification for this.

The following example assumes you are trying to create an admin view that lists all products a specific tenant has in their store.

`app\Http\Controllers\TenantProducts.php`
```php
public function index(Request $request)
{
  // We currently do not have a tenant identified as this is the "admin sub-site"

  // Get our tenant (A Customer)
  $customer = App\Models\Customer::firstOrFail($request->get('tenant_id'));
  // Tell tenancy, that we want to access the above tenant

>   Tenancy::setTenant($customer)

Reviewer 2: Would recommend using the `ConnectionResolver` in order to set the connection, rather than completely setting the tenant as per performance, security etc.

  // We are still in the "admin sub-site", but we now also have an active tenant
  // And we can access the information stored in their database.

  // We will assume that Products uses the `OnTenant` Trait
  $tenant_products = Products::all();

  // Return the products that belong to the tenant still within our "admin sub-site"
  return $tenant_products

}
```

## Considerations to take into account
1. **Storing users in tenant databases**
When the User Model has the `OnTenant` trait applied to it, you will not be able to have those users login to any "non-tenant" routes. Because there is no tenant identified, it will not be possible to authenticate the user.

> To overcome this you will need to either create two different User models (Admin & User), or will need to store all users in the system database. The option you choose will depend on youre goals and objectives.

Reviewer: This is not true and reminds me a lot about the restrictions that hyn/multi-tenant had. If you would want an admin dashboard that allows authenticating users from a tenant, you can create a log in page that asks for their credentials and one more field with a slug, unique hash or something else to identify the tenant with. Then, modify the authentication layer to see whether the identifier is in the payload and set up a tenant database connection on the fly..

t/t is built to be as flexible as possible, the only limitation is your own imagination.

2. **Subdomain identification**
When using subdomain identification, it is best practice to preform an additional check when registering a new tenant that they do not try to use a subdomain you have configured for a "non-tenant site".

> I.E. Do not allow users to register the subdomains (`www`, `admin`, etc...)

Reviewer: Without providing an actual implementation for this, I'm not sure this advice is appreciated. Even mentioning this might sound condescending because it seems pretty obvious you want this.
